libia2/include/ia2.h

```diff
@@ -210,23 +208,41 @@ static int insecure_pkey_mprotect(void *ptr, size_t len, int prot, int pkey) {
         "xor %%ecx,%%ecx\n"                                                    \
         "# eax = old pkru; also zeroes edx, which is required for wrpkru\n"    \
         IA2_RDPKRU "\n"                                                        \
-        "# save pkru in r12d\n"                                                \
+        "# save pkru in r12d. XXX: if a callee here spills r12, it could\n"    \
+        "# be corrupted by another thread subsequently corrupt the pkru\n"     \
+        "# when we restore it from r12d\n"                                     \
```

Comment:
Can we just assert the initial pkey value is compartment 0 and restore back to that value when finished? I don't like the idea of restoring from something that will likely be spilled.

Reply:
We can't--in new threads, we need to restore the pkey that the thread started with, which is whatever the creating compartment had, not necessarily zero. I don't think this is ultimately the right way to do this either (given that we don't control register allocation of all C code that we want to call here), but I'm not sure what is.

Reply:
Filed as #257.